[Auditbeat] Process: Add hash of executable

cwurm opened the pull request:
Adds the hash(es) of the process executable to `process.hash.*`. The default is to add SHA-1 only as `process.hash.sha1`.

## Design
Re-using the hashing code from `file_integrity` (see next section for some of the copied places) introduces a `FileHasher` type in a new package `auditbeat/helper/hasher`. Configured using its own `Config` and created using `NewFileHasher(config, done)`, it has only one other useful method, `HashFile(path)` that does the hashing incl. throttling.

In the future, we could re-use this hasher in other places (e.g. `auditd.paths`, Winlogbeat's `process.[parent.]executable`, the `add_process_metadata` processor).

Executable hashing can be quite expensive. When hashing all execs of all about 450 running processes on my macOS it takes about 10 seconds. That's why I moved the hashing logic to an `enrichProcess()` function that is only called when a new process is discovered. When the dataset is run for the first time, it will take a bit, but afterwards, it should be much better.

## Code
The code is mostly copied from the `file_integrity` module:
- Hashing https://github.com/elastic/beats/blob/bb3f1168f55efdfe92ee6aceb2893162656994ae/auditbeat/module/file_integrity/event.go#L357
- Configuration https://github.com/elastic/beats/blob/bb3f1168f55efdfe92ee6aceb2893162656994ae/auditbeat/module/file_integrity/config.go#L32-L33
- Throttling https://github.com/elastic/beats/blob/bb3f1168f55efdfe92ee6aceb2893162656994ae/auditbeat/module/file_integrity/scanner.go#L178

## Configuration
3 new parameters, same as for `file_integrity` just with a `process.hash` prefix:
- `process.hash.hash_types` (default `[sha1]`)
- `process.hash.max_file_size` (default `100 MiB`)
- `process.hash.scan_rate_per_sec` (default `50 MiB`)

## Follow up
Two immediate ones:
1. Refactor the `file_integrity` module to use the new `FileHasher`. I tried doing this in this PR, but it became quite complex surprisingly quickly. The hashing logic is strewn across various files and types.
2. Introduce a `hash_max_delay` (naming tbd) parameter to configure how long to wait for a hashing to happen (based on `scan_rate_per_sec` which we honor here even for regular process events, something we might want to do in `file_integrity` as well). At the moment, it seems the hashing time is essentially unbounded - there is `max_file_size`, but given enough files hashing could take over CPU and I/O.

@andrewkroh @adriansr - adding you both as reviewers since you're both familiar with the hashing code from `file_integrity`.

adriansr commented on auditbeat/helper/hasher/hasher.go:
```diff
@@ -0,0 +1,285 @@
+	SHA3_384:    {},
+	SHA3_512:    {},
+	XXH64:       {},
+}
```
You can make this map store the constructors: `map[HashType]func()hash.Hash`. 

This helps avoid the big switch statement in HashFile below. Also makes defining the separate HashType constants less necessary, so all can be kept in just one place.

cwurm replied:
I think this would work if the constructors of all hashes would conform to `func() hash.Hash`. Unfortunately, not all of them do, e.g. blake2b is `func New256(key []byte) (hash.Hash, error)`.

adriansr replied:
In that case you can do something like:
```
BLAKE2B_256: func() hash.Hash {
    h,_ := blake2b.New256(nil)
    return h
}
```

But again, this may be out of scope for this PR.

cwurm replied:
I like it, let me do that. Thanks!

adriansr commented on CHANGELOG.next.asciidoc:
```diff
@@ -104,6 +104,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Auditd module: Add `event.outcome` and `event.type` for ECS. {pull}11432[11432]
 - Package: Enable suse. {pull}11634[11634]
 - Add support to the system package dataset for the SUSE OS family. {pull}11634[11634]
+- Process: Hash executable. {pull}11722[11722]
```
I think this is too cryptic. Can you make it more descriptive? "Support hashing the process executable" or similar.

cwurm replied:
I've changed to `Process: Add file hash of process executable.`.

adriansr commented on auditbeat/helper/hasher/hasher.go:
```diff
@@ -0,0 +1,285 @@
+
+	for _, ht := range c.HashTypes {
+		if !ht.IsValid() {
+			errs = append(errs, errors.Errorf("invalid hash_types value '%v'", ht))
```
Can we make it more error tolerant by lowercasing user input so Sha256 or SHA256 don't cause an error?

cwurm replied:
Good idea, done. Very easy to do thanks to go-ucfg.

adriansr commented:
jenkins, test this.

cwurm commented:
1. Rebased on master to resolve conflicts.
2. Simplified hash construction - thanks @adriansr!
3. Moved detailed hash config to auditbeat.reference.yml - same as for `file_integrity`.

andrewkroh commented on auditbeat/helper/hasher/hasher.go:
```diff
@@ -0,0 +1,264 @@
+
+// Config contains the configuration of a FileHasher.
+type Config struct {
+	HashTypes           []HashType `config:"hash_types,replace"`
```
q: What's does the `,replace` option do? I haven't see that used before.

cwurm replied:
`replace` allows `process.hash.hash_types` to be empty and disable hashing altogether that way (e.g. if there is a performance problem, or simply to restore current behavior). Only when the config option is not present at all does the default (`sha1`) take effect.

Without `replace`, the default value will be used even when the option is present but empty, so there would be no way of disabling it.

I would prefer something called `allow_empty` or similar, but `replace` is what I found works as well.

@urso maybe you can take a quick look at whether this is the right use.

willemdh commented:
@cwurm 

What kind of hashes will the functionality allow to create? Any chance and imphash is on the roadmap? 

"The imphash value can be used to identify malware samples that are related to one another. It can also be utilized to identify similar samples created by a certain threat group."

Maybe not ideal for high granluarity process auditing, but combine this with the file_integrity module and you could kind of create your own antivirus with TotalVirus imphash lookups in a Logstash pipeline

https://www.virustotal.com/intelligence/help/

![image](https://user-images.githubusercontent.com/6462991/57507881-4a953f00-7300-11e9-8526-e66c5fabc425.png)
